### ISSUE TITLE
NFS with gcePersistentDisk mount not working with GKE / k8s

### ISSUE THREAD

**Issue by AndryBray** [kubernetes/kubernetes#34252]
I premise I'm opening a ticket only after almost 48 hours of searches & tests.
I tested even your official NFS examples and unofficial docker images and configs - nothing works.

**Kubernetes version**:
```
Client Version: version.Info{Major:"1", Minor:"3", GitVersion:"v1.3.7", GitCommit:"a2cba278cba1f6881bb0a7704d9cac6fca6ed435", GitTreeState:"clean", BuildDate:"2016-09-12T23:15:30Z", GoVersion:"go1.6.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0", GitCommit:"a16c0a7f71a6f93c7e0f222d961f4675cd97a46b", GitTreeState:"clean", BuildDate:"2016-09-26T18:10:32Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- Google Cloud - Container Engine Cluster (europe-west1-b	, 1 node, 1 vCPU, RAM 1.7 GB, node ver. 1.4.0 ) - All permissions on (full/enabled/read-write)
- Every command & test is running through Google Cloud Shell VM.

**The Problem:**
I'm trying to run the same environment (which was working on previous versions of k8s) on Google Container Engine with this config:
**1 deployment (1 replica container) as NFS server** connected directly to Google Persistent Disk (Standard - 10Gb) - gcr.io/google_containers/volume-nfs image
**1 deployment (1 or 2 replicas / containers) as web server** - php:7.0.8-apache official docker image

**The main scope is to share 1 pd-volume between each webapp-server replica** but when I create my deployments, during the web-app deployment creation, it breaks with NFS/mount problem.
(the NFS server starts even if with some warning)

**Here some other info about the mount error:**

`kubectl describe po web-app`
```
....[omitted]...
Mounting arguments: 10.0.2.9:/exports /var/lib/kubelet/pods/...../volumes/kubernetes.io~nfs/pv-nfs-webapp-logs nfs []
Output: mount: wrong fs type, bad option, bad superblock on 10.0.2.9:/exports,
       missing codepage or helper program, or other error
       (for several filesystems (e.g. nfs, cifs) you might
       need a /sbin/mount.<type> helper program)

       In some cases useful info is found in syslog - try
       dmesg | tail or so.

....[omitted]...
```

From this moment `10.0.2.9` is the NFS Server host IP

If I connect inside the web-app pod I can ping the NFS server (by IP), and it's sharing the right folders as described:
`showmount -e 10.0.2.9 `
```
Export list for 10.0.2.9:
/        *
/exports *
```

The nmap report confirms running ports are OK:
```
Nmap scan report for 10.0.2.9
Host is up (0.000026s latency).
Not shown: 998 closed ports
PORT     STATE SERVICE
111/tcp  open  rpcbind
2049/tcp open  nfs
```

**Here the logs of NFS pod:**
`kubectl logs nfs-deployment`
```
Serving /exports
Serving /
rpcinfo: can't contact rpcbind: : RPC: Unable to receive; errno = Connection refused
Starting rpcbind
exportfs: / does not support NFS export
NFS started
```

`kubectl describe po nfs-deployment`

```
Name:           nfs-deployment-...
Status:         Running
IP:             10.0.2.9
...
Controllers:    ReplicaSet/nfs-deployment-115940229
Containers:
  nfs-server:
    Container ID:       docker://.....
    Image:              gcr.io/google_containers/volume-nfs
    Image ID:           docker://sha256:...
    Ports:              2049/TCP, 111/TCP
    Requests:
      cpu:                      10m
    State:                      Running
      Started:                  Thu, 06 Oct 2016 16:45:40 +0200
    Ready:                      True
    Restart Count:              0
    Environment Variables:      <none>
Conditions:
  Type          Status
  Initialized   True 
  Ready         True 
  PodScheduled  True 
Volumes:
  nfs-pd:
    Type:       GCEPersistentDisk (a Persistent Disk resource in Google Compute Engine)
    PDName:     disk-1
    FSType:     ext4
    Partition:  0
    ReadOnly:   false
  default-token-f7r08:
    Type:       Secret (a volume populated by a Secret)
    SecretName: default-token-f...
```

I tested many configurations but the same working on older k8s version is like this:
(in order of creation)

**nfs.yaml**
```
# NFS POD / SERVER (PERSISTEN DISK SHARING)
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: nfs-deployment
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: nfs-server
        role: backend
        #env: staging
    spec:
      restartPolicy: Always
      containers:
      - name: nfs-server
        #image: gcr.io/google-samples/nfs-server:1.1
        image: gcr.io/google_containers/volume-nfs
        #imagePullPolicy: Always
        ports:
          #- name: nfs
            # containerPort: 2049
            # hostPort: 2049
          - name: nfs
            hostPort: 2049
            containerPort: 2049
            protocol: TCP
          - name: mountd
            hostPort: 20048
            containerPort: 20048
            protocol: UDP
          - name: rpcbind
            hostPort: 111
            containerPort: 111
        volumeMounts:
          - mountPath: "/exports"
            name: nfs-pd
        securityContext:
            privileged: true
        resources:
          requests:
            cpu: 10m
      volumes:
        - name: nfs-pd
          gcePersistentDisk:
            pdName: disk-1
            fsType: ext4
```

**pvc.yaml**

```
apiVersion: v1
kind: PersistentVolume
metadata:
  name: pv-nfs-webapp
spec:
  capacity:
    storage: 5Gi
  accessModes:
    - ReadWriteMany
  persistentVolumeReclaimPolicy: Recycle
  nfs:
    #this will be changed after nfs server is ready
    server: 10.0.2.7
    path: "/exports"
    readOnly: false
---
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: pvc-nfs-webapp
spec:
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 5Gi
```

**web.yaml**
```
# POD DEPLOYMENT
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: webapp-deployment
spec:
  #replicas: 1
  template:
    metadata:
      labels:
        app: webapp
        role: frontend
    spec:
      restartPolicy: Always
      containers:
      - name: webapp
        image: gcr.io/my-app-reg/my-webapp:v0.0.1
        ports:
          - name: apache-http
            containerPort: 80
          - name: apache-https
            containerPort: 443
        resources:
          requests:
            cpu: 10m
        livenessProbe:
          httpGet:
            path: /kubernetes-api/status/healthz
            port: 80
            scheme: HTTP
          initialDelaySeconds: 20
          periodSeconds: 60
          timeoutSeconds: 10
        readinessProbe:
          httpGet:
            path: /kubernetes-api/status/readiness
            port: 80
            scheme: HTTP
          initialDelaySeconds: 15
          timeoutSeconds: 10
        volumeMounts:
          - name: nfs-vol-webapp
          mountPath: /var/www/html/
      volumes:
        - name: nfs-vol-webapp
          persistentVolumeClaim:
            claimName: pvc-nfs-webapp
```

**Comment by AndryBray** [comment]
If I comment volumes and volumeMounts from my webapp.yaml deployment, the pod runs without problems.
So I connect to its bash just to try to mount manually the NFS:

`mount  10.0.0.8:/exports /mnt/ -vvv -o nolock`
``` 
mount.nfs: timeout set for Fri Oct  7 07:11:13 2016
mount.nfs: trying text-based options 'nolock,vers=4,addr=10.0.0.8,clientaddr=10.0.0.9'
mount.nfs: mount(2): Operation not permitted
mount.nfs: trying text-based options 'nolock,addr=10.0.0.8'
mount.nfs: prog 100003, trying vers=3, prot=6
mount.nfs: trying 10.0.0.8 prog 100003 vers 3 prot TCP port 2049
mount.nfs: prog 100005, trying vers=3, prot=17
mount.nfs: trying 10.0.0.8 prog 100005 vers 3 prot UDP port 20048
mount.nfs: mount(2): Operation not permitted
mount.nfs: Operation not permitted
```

Tested even:
`mount  -t nfs 10.0.0.8:/exports /mnt/`
`mount  -t nfs 10.0.0.8:/exports /mnt/ -o vers=4`
`mount  -t nfs 10.0.0.8:/exports /mnt/ -o nolock,vers=4`
etc.etc.


Do I miss something in my config? Maybe firewall rules? DNS?

**Comment by jingxu97** [comment]
@AndryBray , could you please share your kubelet log with us? If you prefer, you can email me via jinxu@google.com or find me at slack @jinxu. Thank you!

**Comment by jingxu97** [comment]
@AndryBray could you also share your project name and cluster name? Thanks!

**Comment by AndryBray** [comment]
@jingxu97 it was just a test project / cluster which I've destroyed.
I'm going to rung again the cluster and in some minute I will write you again

**Comment by AndryBray** [comment]
@jingxu97 I've just sent you an email with details about cluster and project, please let me know if you need other infos.
Thank you for your support

**Comment by AndryBray** [comment]
@jingxu97 the cluster has been stopped because it was just for testing reason.
We are blocked with NFS file sharing issue...
Please could you confirm me if you received the email? Thank you

**Comment by jingxu97** [comment]
@AndryBray , recently we changed the default VM image to GCI which is currently lacking NFS client toolings. We are in the process of adding them in the GCI. In the mean time, the following the bottom of the link https://cloud.google.com/container-engine/docs/node-image-migration to change the image back to Debian 7-based container. 

Sorry for this inconvenience and please let us know if you have any issue. 
@saad-ali , could you please also share updates for this nfs issue?

**Comment by saad-ali** [comment]
You can follow the progress of adding in Storage client binaries to GCI in https://github.com/kubernetes/kubernetes/issues/34127

**Comment by AndryBray** [comment]
@jingxu97 the container and persistent disk zone was "europe-west1-b"

**Comment by jingxu97** [comment]
@AndryBray since the cluster is stopped, I cannot access the log now. But based on the description, we identified the problem with GCI image. Could you try the link  or the PR @saad-ali just sent out. Please let us know if you still have any issue. Thanks!

**Comment by AndryBray** [comment]
@jingxu97 If you need me to start the cluster I can do it.

Do you know approximately when the GCI issue will be solved?

We'd like to avoid using old images as we are starting right now with a new project so we can wait for a while (of course if you plan to fix it asap) for the bugfix

**Comment by jingxu97** [comment]
@saad-ali , is there an expected date for the GCI fix?

**Comment by saad-ali** [comment]
We are preliminarily targeting a v1.4.x release to include NFS and GlusterFS binaries in GCI. In the meantime, as @jingxu97 mentioned, those that dependent on the existence of these client binaries in the OS image can switch back to the old (non-default) `container-vm` image (see the bottom of [this page](https://cloud.google.com/container-engine/docs/node-image-migration)).

**Comment by saad-ali** [comment]
Closing this bug as a duplicate of https://github.com/kubernetes/kubernetes/issues/33447
